src/graphics/device.js

```diff
@@ -2812,6 +2840,200 @@ Object.assign(pc, function () {
+            var gl = this.gl;
+
+            var glShader = isVertexShader ? this.vertexShaderCache[src] : this.fragmentShaderCache[src];
+
+            if (!glShader) {
+                // #ifdef PROFILER
+                var startTime = pc.now();
+                this.fire('shader:compile:start', {
+                    timestamp: startTime,
+                    target: this
+                });
+                // #endif
+
+                glShader = gl.createShader(isVertexShader ? gl.VERTEX_SHADER : gl.FRAGMENT_SHADER);
+
+                gl.shaderSource(glShader, src);
+                gl.compileShader(glShader);
+
+                // #ifdef PROFILER
+                var endTime = pc.now();
+                this.fire('shader:compile:end', {
+                    timestamp: endTime,
+                    target: this
+                });
+                this._shaderStats.compileTime += endTime - startTime;
+                // #endif
+
+                if (isVertexShader) {
+                    this.vertexShaderCache[src] = glShader;
+                    // #ifdef PROFILER
+                    this._shaderStats.vsCompiled++;
+                    // #endif
+                } else {
+                    this.fragmentShaderCache[src] = glShader;
+                    // #ifdef PROFILER
+                    this._shaderStats.fsCompiled++;
+                    // #endif
+                }
+            }
+
+            return glShader;
+        },
+
+        compileAndLinkShader: function (shader) {
+            var gl = this.gl;
+
+            var definition = shader.definition;
+            var glVertexShader = this.compileShaderSource(definition.vshader, true);
+            var glFragmentShader = this.compileShaderSource(definition.fshader, false);
+
+            var glProgram = gl.createProgram();
+
+            gl.attachShader(glProgram, glVertexShader);
+            gl.attachShader(glProgram, glFragmentShader);
+
+            if (this.webgl2 && definition.useTransformFeedback) {
+                // Collect all "out_" attributes and use them for output
+                var attrs = definition.attributes;
+                var outNames = [];
+                for (var attr in attrs) {
+                    if (attrs.hasOwnProperty(attr)) {
+                        outNames.push("out_" + attr);
+                    }
+                }
+                gl.transformFeedbackVaryings(glProgram, outNames, gl.INTERLEAVED_ATTRIBS);
+            }
+
+            gl.linkProgram(glProgram);
+
+            // Cache the WebGL objects on the shader
+            shader._glVertexShader = glVertexShader;
+            shader._glFragmentShader = glFragmentShader;
+            shader._glProgram = glProgram;
+
+            // #ifdef PROFILER
+            if (definition.tag === pc.SHADERTAG_MATERIAL) {
+                this._shaderStats.materialShaders++;
+            }
+            // #endif
+        },
+
+        createShader: function (shader) {
```

Comment:
slightly confusing terminology. In constructor of `pc.Shader` we call `device.createShader(this)` which compiles the shader?

Reply:
Yep, it's not ideal. It's tricky since a pc.Shader is a WebGL program and a WebGL shader doesn't really have an equivalent in PlayCanvas. This is internal to not a big deal for now. But we could refactor a bit further later.